ev3dev2/motor.py

```diff
@@ -2182,6 +2189,138 @@ def turn_left(self, speed, degrees, brake=True, block=True):
         """
         self._turn(speed, abs(degrees) * -1, brake, block)
 
+    def odometry_coordinates_log(self):
+        log.info("%s: odometry angle %s at (%d, %d)" %
+            (self, math.degrees(self.theta), self.x_pos_mm, self.y_pos_mm))
+
+    def odometry_start(self, theta_degrees_start=0.0, x_pos_start=0.0, y_pos_start=0.0):
+        """
+        Ported from:
+        http://seattlerobotics.org/encoder/200610/Article3/IMU%20Odometry,%20by%20David%20Anderson.htm
+
+        A thread is started that will run util the user calls odometry_stop()
+        which will set odometry_thread_run to False
+        """
+
+        def _odometry_monitor():
+            left_previous = 0
+            right_previous = 0
+            self.theta = math.radians(theta_degrees_start)  # robot heading
+            self.x_pos_mm = x_pos_start  # robot X position in mm
+            self.y_pos_mm = y_pos_start  # robot Y position in mm
+            SLEEP_TIME = 0.05  # 50ms
+            TWO_PI = 2 * math.pi
+
+            while self.odometry_thread_run:
+
+                # sample the left and right encoder counts as close together
+                # in time as possible
+                left_current = self.left_motor.position
+                right_current = self.right_motor.position
+
+                # determine how many ticks since our last sampling
+                left_ticks = left_current - left_previous
+                right_ticks = right_current - right_previous
+
+                # Have we moved?
+                if not left_ticks and not right_ticks:
+                    time.sleep(SLEEP_TIME)
+                    continue
+
+                # log.debug("%s: left_ticks %s (from %s to %s)" %
+                #     (self, left_ticks, left_previous, left_current))
+                # log.debug("%s: right_ticks %s (from %s to %s)" %
+                #     (self, right_ticks, right_previous, right_current))
+
+                # update _previous for next time
+                left_previous = left_current
+                right_previous = right_current
+
+                # rotations = distance_mm/self.wheel.circumference_mm
+                left_rotations = float(left_ticks / self.left_motor.count_per_rot)
+                right_rotations = float(right_ticks / self.right_motor.count_per_rot)
+
+                # convert longs to floats and ticks to mm
+                left_mm = float(left_rotations * self.wheel.circumference_mm)
+                right_mm = float(right_rotations * self.wheel.circumference_mm)
+
+                # calculate distance we have traveled since last sampling
+                mm = (left_mm + right_mm) / 2.0
+
+                # accumulate total rotation around our center
+                self.theta += (left_mm - right_mm) / self.wheel_distance_mm
+
+                # and clip the rotation to plus or minus 360 degrees
+                self.theta -= float(int(self.theta/TWO_PI) * TWO_PI)
```

Comment:
This seems strange... If we to above 2pi it's shifted down to be greater than 0, but we can go below zero until we get to -2pi... In other words, we allow a range of `[-2pi,2pi]` which is 2 rotations. Shouldn't this me a mod here to give us `(0,2pi]`?

Reply:
I'm not sure...this is a straight port of the C implementation I found

Reply:
Once I get further along with this I can test both methods and see which one looks correct

Reply:
> For the angle, I'd suggest we maintain 90 being to the right (positive X) so that triggered functions work as they would on a typical unit circle layout

To clarify here, if the robot drives forward in a straight line `self.theta` will be 0 because that is the initial value of `self.theta`. Are you saying it is correct at 0 or it should be 90?  If we change the intiial value to 90 we will need to tweak some of the other match in this loop (I had tried this and using 90 as the initial value causes x to increase while y remains 0 when driving in a straight line).

One option here would be to store two values:
- `self.theta` can remain as is, this will be a measure of how many degrees the robot has turned
- A new variable `self.angle_direction` (something like that) that would start at 90 and would tell you which way the robot is facing

Reply:
If we want to make that change, it should just be a matter of changing the initial angle to 90 degrees and then swapping sin with cos and vice versa.

Reply:
If we want to make that change, it should just be a matter of changing the initial angle to 90 degrees and then swapping sin with cos and vice versa.

Reply:
I fixed all of this...we start at 90 degrees now.